coordinator-common/src/test/java/org/apache/kafka/coordinator/common/runtime/HdrHistogramTest.java

```diff
@@ -172,4 +177,39 @@ public void testHistogramDataReset() {
         assertEquals(numEventsInFirstCycle, hdrHistogram.count(now + maxSnapshotAgeMs));
         assertEquals(numEventsInSecondCycle, hdrHistogram.count(now + 1 + maxSnapshotAgeMs));
     }
+
+    @Test
+    public void testLatestHistogramRace() throws InterruptedException, ExecutionException {
+        long maxSnapshotAgeMs = 10L;
+        long now = System.currentTimeMillis();
+        HdrHistogram hdrHistogram = new HdrHistogram(maxSnapshotAgeMs, MAX_VALUE, 1);
+        ExecutorService countExecutor = Executors.newFixedThreadPool(2);
+        for (int i = 1; i < 10000; i++) {
+            int numEvents = 2;
+            for (int j = 0; j < numEvents; j++) {
+                hdrHistogram.record(i);
+            }
+            final long moreThanMaxAge = now + maxSnapshotAgeMs + 1;
+            now = moreThanMaxAge;
+            CountDownLatch latch = new CountDownLatch(1);
+            Callable<Long> countTask = () -> {
+                try {
+                    assertTrue(latch.await(500, TimeUnit.MILLISECONDS));
+                    return hdrHistogram.count(moreThanMaxAge);
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+            };
+            Future<Long> t1Future = countExecutor.submit(countTask);
+            Future<Long> t2Future = countExecutor.submit(countTask);
+            latch.countDown();
+            long t1Count = t1Future.get();
+            long t2Count = t2Future.get();
+            assertTrue(
+                numEvents == t1Count && numEvents == t2Count,
```

Comment:
actually, is this what we want? my understanding is that if we have two threads concurrently calling latestHistogram, we want the first thread to reset the recorder and return the previous histogram where the other thread should return the reset-ed histogram.

i may be wrong but both thread counts returning 2 doesn't seem right

Reply:
> actually, is this what we want? my understanding is that if we have two threads concurrently calling latestHistogram, we want the first thread to reset the recorder and return the previous histogram where the other thread should return the reset-ed histogram.

Yes, this is what we want for this particular test. We want the threads to race for the snapshot reset, we want the thread that wins the race to be guaranteed to reset the snapshot, and we want the thread that loses the race to be guaranteed to accept the reset snapshot and not to reset the snapshot itself.

> i may be wrong but both thread counts returning 2 doesn't seem right

It is actually the correct behavior given the current API. We read from both threads using the same timestamp, which is why the latest condition I mentioned above is guaranteed - that the thread that loses the race will not reset the snapshot itself.

I think some confusion might be arising from the fact that the histogram relies user thread collaboration for the snapshot freshness (e.g. like Guava Cache implementations rely on user thread collaboration for entry expiration). So the actual max age of the histogram snapshot in case no reads are performed can be infinity.
- We don't really want to do that maintenance on the write path, and while it's possible to do it separately and not rely on user threads (e.g. the same way later versions of Caffeine, a Guava Cache successor, allow specifying an executor/scheduler for entry expiration) I don't think this is a necessary feature for the wrapper.

Reply:
ah, i understand where i got confused - with this PR a call to `latestHistogram()` will always return a MAX_AGE_SNAPSHOT worth of data (as it should). thanks for the explanation!